Ica Automation Scripts

jwarnn opened the pull request:
I've made two scripts that are to automate starting analyses on ICA; both are quite different because of the way the sample data is getting to ICA and the pipelines that they are based off of.  I have both running on screens on my account on the linux workstation; just to make sure that they are validated and work. Feel free to ask a lot of questions because you guys might not under the reasoning around ICA steps. I also would like to write scripts that download the files that are needed for reporting purposes. Thanks.

jwarnn committed:
Create auto_mycosnp_ICA.py

jwarnn committed:
Create auto_Granduer_ICA.py

jwarnn committed:
Create README.md

jwarnn committed:
Update auto_Granduer_ICA.py

jwarnn committed:
Update auto_mycosnp_ICA.py

erinyoung commented on ICA/README.md:
```diff
@@ -0,0 +1,13 @@
+
+# Known Issues
+1. Sometimes analyses fail on ICA and I need to work out the best way to restart these runs. This most likly intells releasing pipelines on ICA and moving each scrip from running in Testing to Production projects; which should only happen after we have some evaulation of these scripts.
+2. Currently no downloading has been automated in the scripts. I need to clarify with bioinformaticains what needs to be downloaded and I think version control would be easier with creating scripts that are different then the ones starting analyses.
```
Can you include this information in the main readme?

We can start a wiki for this repo if you think this would be better located behind a couple of clicks.

jwarnn replied:
Yeah I will just scrap this readme and make ICA Automation Scripts a subheader with the rest of the readme below.

erinyoung commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,75 @@
+
+        # Create icav2 argument
+        arg = "icav2 projectpipelines start nextflow UPHL-BioNGS_Granduer_V2 --input reads:%s " % ica_id
+        arg = arg + "--input project_dirs:fol.fb0091ef9e4d4e515b6408da63045a75,fol.90b6b7865bc7434384e408da6377c84,fol.8802646a4148485884df08da6377c840,fol.17212ac619d44b68296008da58d44868,fol.ae742e54d4a54a8acb9c08da64f9d308,fol.97c3321c555a4b1c672f08da58d44868 "
```
Are these always going to be the directory names?

They seem kind of long and weird to be static.

jwarnn replied:
Yes the files have a static ICA id, fol.fb0091ef9e4d4e515b6408da63045a75; the CLI tool icav2 uses these ids instead of path names for the command ```icav2 projectpipelines start nextflow``` this might change with updates to that tool. IDs will not change unless it is deleted which should not be needed for a pipeline that is up and working; although they would change for the next version.
This is probably a good place to bring up versioning. I didn't think about that when I made these scripts but this script will be for Granduer_V2 or Granduer 2.0.20220610; when we need to update to a new version I would upload the new files and they would have new ids and I would just update those two lines; 64 and 65.

erinyoung replied:
I think a static grandeur version is fine. Updating the version is going to require a new clia validation anyway.

erinyoung commented on ICA/auto_mycosnp_ICA.py:
```diff
@@ -0,0 +1,190 @@
+    runs_completed_analysis=os.listdir("/Volumes/IDGenomics_NAS/ICA/Runs_On_ICA")
+#This maybe unneeded but during testing I think the os package generates this file that needs to be removed from list
+    try:
+        runs_completed_analysis.remove('.DS_Store')
```
It's best to keep the .DS_Store and @eaDir in case someone on their mac is looking in this directory

erinyoung commented on ICA/auto_mycosnp_ICA.py:
```diff
@@ -0,0 +1,190 @@
+    for i in range(len(tmp)):
+        runs_on_ica.append(tmp[i].split()[0].split('/')[2])
+# The plan is to download only the files that are commonly needed by Tom and use these directories as a way to way to moniter if anaylsis has completed.
+    runs_completed_analysis=os.listdir("/Volumes/IDGenomics_NAS/ICA/Runs_On_ICA")
```
Can you add a glob that could be something like
```
import glob
runs_completed_analysis=glob("/Volumes/IDGenomics_NAS/ICA/Runs_On_ICA/UT*")
```

jwarnn replied:
Do we want all Granduer runs to run on ICA or the linux work station? I was thinking we could change General_LW_scripts/MiSeq_runs_basespace.sh to not run Granduer on the work station once we are confident in ICA.

erinyoung replied:
Once CLIA has been approved to run Grandeur on ICA, all of them will be run on ICA. It has to be approved by our CLIA officer before we make the switch, though.

erinyoung replied:
So, yes, the goal of this script would be to replace `MiSeq_runs_basespace.sh`

jwarnn replied:
Do you want to download the whole Granduer analysis folder when its completed or just certain parts? What would be best for you?

erinyoung replied:
The entire grandeur directory would be ideal. 

I don't mind if it's compressed before download.

tives82 replied:
Here are the mycosnp_ICA results files I'll need downloaded. 

/results/stats/qc_report/qc_report.txt
/results/multiqc/multiqc_report.html
/results/combined/vcf-to-fasta/vcf-to-fasta.fasta
/results/combined/snpdists/combined.tsv

erinyoung replied:
@tives82 , you need the fastq files, too, right?

jwarnn replied:
For the CLIA validation some of those folders were 100 GB and took about 2 hrs to download. I am reaching out to ICA to see if I can compress those files on there platform and then download them.  They will probably tell me to use bench but thats costs compare to the just dealing with the time that I don't think it would be worth it.

tives82 replied:
@erinyoung, Yeah, I'll need the fastq files downloaded as well. 
@jwarnn, I know snpdists wasn't one of the files we had talked about previously but I figure it's a good idea to have the snpdists file be downloaded due to more Epis requesting it and the ability for us to upload it into Microreact.

erinyoung commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,75 @@
+    #This  block collects the information of data and runs on ICA
+    bashCommand = "icav2 projectdata list --file-name UT-M03999-  --parent-folder / --match-mode FUZZY"
+    tmp= icav_out(bashCommand)
+    bashCommand = "icav2 projectdata list --file-name UT-M07101-  --parent-folder / --match-mode FUZZY"
```
We actually have three MiSeqs (and one iSeq, but I don't think that's going to be used anytime soon)

jwarnn replied:
I should probably add those too just incase and maybe try to make it a single request to ICA.

erinyoung commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,75 @@
+        except:
+            del tmp[i]
+            continue
+        if date < cutoff:
```
Aren't all the dates greater than the cutoff now?

jwarnn replied:
Yes, I have the script running and would like it to kick off just for new runs. No need to run Granduer twice.

erinyoung replied:
I think that's a good idea. Because it would start exploring all the old runs and take forever. 

What if the cutoff was something like 

```
from datetime import date, datetime, timedelta

delta = 10
today = datetime.now()    
recent = today - timedelta(days=delta)

cutoff=datetime.strptime(recent, '%y%m%d')
```

jwarnn replied:
I like this only needs slight modification. Recent is already a datetime object; so last line is unneeded.

erinyoung commented on ICA/auto_mycosnp_ICA.py:
```diff
@@ -0,0 +1,190 @@
+    tmp=[x.replace('\t', ' ') for x in tmp]
+    return tmp
+
+# While loop is infinate will keep script running as long as it dosent error out.
```
I think there is a typo in this comment (infinite != infinate)

erinyoung commented on ICA/auto_mycosnp_ICA.py:
```diff
@@ -0,0 +1,190 @@
+            bashCommand = "icav2 projectdata list  --parent-folder  " +  run + "output/fastqs/Samples/Lane_1/"
+            tmp= icav_out(bashCommand)
+            for j in tmp:
+                ica_names.append(j.split(" ")[0])
```
I think you can add your "Undetermined" check here

```
for j in tmp:
     if not "Undetermined" in j:
     ica_names.append(j.split(" ")[0])
```

jwarnn replied:
Adding this ```if not "Undetermined" in j.split(" ")[0]:```

jwarnn committed:
Update README.md to include ICA info

jwarnn committed:
Delete README.md

erinyoung commented on README.md:
```diff
@@ -19,3 +19,17 @@ MiSeq_runs_basespace.sh
+These scripts are intened to make the running of analyses on ICA for production purposes automated as well as downloading the most common files that need to be reviewed by human eyes. Each is written in python and uses the icav2 command line tool provided by Illumina. These are intened to run on screens on the linux work station but whereever they are ran icav2 must already be installed and confi ...
+
+### mycosnp
+Currently this pipeline is only ran on C. auris sampels. These samples are usually uploaded directly by the Illumia instrument after BSSH calls the reads into ICA.
```
I think there is a typo on this line (samples != sampels)

jwarnn committed:
Update auto_Granduer_ICA.py

Adding all miseqs and iseq into a single line and request to ICA.

erinyoung commented on README.md:
```diff
@@ -19,3 +19,17 @@ MiSeq_runs_basespace.sh
+The automation script runs each time these file make there way to ICA. It produces the samplesheet, including control samples, needed for the analysis and uploads it to ICA. It then starts the anaylsis
+
+### Grandeur
+We currently upload sequncing runs that use granduer using the General_LW_scripts/MiSeq_runs_basespace.sh. This script looks for those uploads then sees if they have already had anaylsis preformed on ICA. If not they start that anaylsis.
```
sequencing != sequncing

jwarnn replied:
README.md should have a spell check; comments too. Come on github.

erinyoung commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,73 @@
+    return tmp
+
+# While loop is infinate will keep script running as long as it dosent error out.
+go=1
+while go == 1:
```
It's because of these lines that I'm expecting this script to replace `MiSeq_runs_basespace.sh`, because it seems like this would always be on in the background looking for new runs.

I'd rather not have an infinite loop inside an infinite loop.

erinyoung commented:
Can you include in the readme
1. How to run your scripts on the production account (I imagine it's only `auto_Granduer_ICA.py` or `auto_mycosnp_ICA.py` or `python3 auto_mycosnp_ICA.py` or something similar.
2. How to do these analyses on the command line if the automatic parts failed and this has to be run manually

jwarnn committed:
Update auto_Granduer_ICA.py

Updating project_dirs:fol.90b6b7865bc7434384e408da6377c84 was missing digit 0

jwarnn committed:
Update auto_Granduer_ICA.py

arg used to start analysis was pointed at wrong folder. Needs to be raw folder under the experiment name folder

tives82 commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,75 @@
+import numpy as np
+from datetime import date,datetime
+
+# This funciton uses the ica CLI tool to return info from ICA
```
typo: function != funciton

jwarnn replied:
What is this a spelling bee!

tives82 commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,75 @@
+    tmp=[x.replace('\t', ' ') for x in tmp]
+    return tmp
+
+# While loop is infinate will keep script running as long as it dosent error out.
```
typo: infinite != infinate
typo: doesn't != dosent

jwarnn replied:
What is this grammer school!

tives82 commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,75 @@
+# While loop is infinate will keep script running as long as it dosent error out.
+go=1
+while go == 1:
+    print('Looking for anaylses that need to start!')
```
typo: analysis != anaylses

jwarnn replied:
Spelled wrong but still needs to be plural!

jwarnn committed:
Update auto_Granduer_ICA.py

Fixed spelling errors and change cutoff from hard date to date range.

jwarnn committed:
Update auto_mycosnp_ICA.py

Added one liner over three code block. Clearer understanding in one liner.

jwarnn committed:
Update README.md

erinyoung commented:
The `work` and `.nextflow` directories won't be needed in either mycosnp or grandeur. Those are just nextflow's temp files.

I was trying to think of why grandeur would have 100G of data for those samples, and that's all I can think of at the moment.

jwarnn commented:
Those weren't included in the download. But 10x larger is really strange and probably means there is some issue. Could it be because of Spades?

jwarnn committed:
Update README.md

jwarnn committed:
Update README.md

cleaning up some more spelling and grammar

jwarnn committed:
Update auto_Granduer_ICA.py

Original script was having trouble with duplicates; improved logic, also increased time between test so that errors in script don't compound.

erinyoung commented on ICA/auto_Granduer_ICA.py:
```diff
@@ -0,0 +1,85 @@
+    tmp=[x.replace('\t', ' ') for x in tmp]
+    return tmp
+
+# While loop is infinite will keep script running as long as it doesn't error out.
+go=1
+while go == 1:
+    print('Looking for analyses that need to start!')
+    #This block set the icav2 tool to correct project in use
+    bashCommand="icav2 projects enter Testing"
+    process = subprocess.Popen(bashCommand.split(" ",3), stdout=subprocess.PIPE)
+    process.communicate()
+
+    #This  block collects the information of data and runs on ICA
```
It's this portion of this file that is holding up this PR. I hesitate to have dual while loops running.

Moving forward we should
1. Remove the loop and have the bash script run this command once the fastq files have finished downloading
2. This loop needs to also search through basespace for new run, download fastq files, and then continue on

I feel bad that this is holding this PR up. Should we remove this file from the PR and at least get the mycosnp stuff going?

erinyoung commented on README.md:
````diff
@@ -19,3 +19,43 @@ MiSeq_runs_basespace.sh
+This is the basic command for starting an analysis with icav2: ```icav2 projectpipelines start nextflow [pipeline id] or [code] [flags] ```. The pipeline is the name of the pipeline, i.e. UPHL-BioNGS_Granduer_V2. Then you must provide the inputs with this syntax ```--input parametercode:dataId,dataId,dataId,... ``` Finally you must provide a name ```--user reference $SEQUENCING_RUN_PIPELINE_Analy ...
+To find data Ids use ```icav2 projectdata list ```.
+##### mycosnp
+Too hard without scripting. The automation script creates an icav2 command; using this command to troubleshoot is probably the best solution. 
````
This is a problem. We don't want to be held hostage by script that only runs in the background.

erinyoung commented on ICA/auto_mycosnp_ICA.py:
```diff
@@ -0,0 +1,187 @@
+                if not "Undetermined" in j.split(" ")[0]:
+                    ica_names.append(j.split(" ")[0])
+
+        ica_ids=[]
+        sample_names=[]
+        for i in ica_names:
+            bashCommand = 'icav2 projectdata list  --parent-folder ' + i
+            tmp= icav_out(bashCommand)
+            for j in tmp:
+                ica_ids.append(j.split(" ")[4]+",")
+                sample_names.append(j.split(" ")[1])
+
+
+        sample_sheet=[]
+        sample_sheet.append('sample_id,fastq_1,fastq_2\n')
+        for i in sample_names:
+            if i.split('_')[3] == 'R1':
+                tmp=i.split('_')[0]+','+i
+                tmp=tmp+','+i.replace('R1','R2')+'\n'
+                sample_sheet.append(tmp)
```
We need the sample sheet creation be a separate script so it can be externally without the loop

erinyoung commented on ICA/auto_mycosnp_ICA.py:
```diff
@@ -0,0 +1,187 @@
+        process.communicate()
+
+        run=list(set(runs_on_ica)-set(runs_completed_analysis))
+        os.mkdir('/Volumes/IDGenomics_NAS/ICA/Runs_On_ICA/%s' % run[0] )
```
Once this is clia validated, it would be awesome if it would to into /Volumes/IDGenomics_NAS/ICA/Runs_On_ICA or a directory for C. auris

erinyoung commented on ICA/auto_mycosnp_ICA.py:
```diff
@@ -0,0 +1,187 @@
+        process = subprocess.Popen(bashCommand.split(" ",3), stdout=subprocess.PIPE)
+        process.communicate()
+
+    # Mycosnp needs examples of clades so these samples need to be included in each run
+        bashCommand = "icav2 projectdata list --parent-folder /Candida_auris_clade_controls/Raw/"
+        controls= icav_out(bashCommand)
+        for i in controls:
+            if i.split('_')[5].split('.')[0] == 'R1':
+                tmp=i.split()[1].split('.')[0]+','+i.split()[1]
+                tmp=tmp+','+i.split()[1].replace('R1','R2')+'\n'
+                sample_sheet.append(tmp)
+            if i.split('_')[5].split('.')[0] == '1':
+                tmp=i.split()[1].split('.')[0]+','+i.split()[1]
+                tmp=tmp+','+i.split()[1].replace('1','2')+'\n'
+                sample_sheet.append(tmp)
```
I think a separate script that does these commands would be useful

erinyoung committed:
Added text to readme

Perhaps now things are even less clear than before.

erinyoung commented:
My apologies for taking so long to get back to this. 

Could we split the mycosnp script into something that can run by humans if need be?

jwarnn committed:
Merge pull request #13 from UPHL-BioNGS/erin-ica-dev

Added text to readme

jwarnn commented:
I converted this into a draft pull request. I need to a little time to make what changes that are suggested. I also want to wait for Illuminia Case#02933186 to be resolved. This help ticket is aimed at eliminating K8s errors while Nextflow scripts are running on ICA which occurs with every Mycosnp run and around 10% of Grandeur runs. Once those two items are resolved I will open this PR back up.

jwarnn commented:
A different approach was taken for these scripts. They do not need to be completely automated as with ICA we have unlimited resources and UPHL used automation to optimized our local resources. The new approach is for a Bioinformatician to know of a run they start a script with that runs name that will download the reads from BSSH, start analysis on ICA, and download results from ICA with keeping the Bioinformatician updated with Slack messages.
